controllers/nodemodulesconfig_reconciler.go

```diff
@@ -0,0 +1,681 @@
+	logger := ctrl.LoggerFrom(ctx)
+
+	nmc := kmmv1beta1.NodeModulesConfig{}
+
+	if err := r.client.Get(ctx, req.NamespacedName, &nmc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Pods are owned by the NMC, so the GC will have deleted them already.
+			// Remove the finalizer if we did not have a chance to do it before NMC deletion.
+			logger.Info("Clearing worker Pod finalizers")
+
+			if err = r.helper.RemoveOrphanFinalizers(ctx, req.Name); err != nil {
+				return reconcile.Result{}, fmt.Errorf("could not clear all Pod finalizers for NMC %s: %v", req.Name, err)
+			}
+
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, fmt.Errorf("could not get NodeModuleState %s: %v", req.NamespacedName, err)
+	}
+
+	if err := r.helper.SyncStatus(ctx, &nmc); err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not reconcile status for NodeModulesConfig %s: %v", nmc.Name, err)
+	}
+
+	// Statuses are now up-to-date.
+
+	statusMap := make(map[string]*kmmv1beta1.NodeModuleStatus, len(nmc.Status.Modules))
+
+	for i := 0; i < len(nmc.Status.Modules); i++ {
+		status := nmc.Status.Modules[i]
+		statusMap[status.Namespace+"/"+status.Name] = &nmc.Status.Modules[i]
+	}
+
+	for _, mod := range nmc.Spec.Modules {
+		moduleNameKey := mod.Namespace + "/" + mod.Name
+
+		logger := logger.WithValues("module", moduleNameKey)
+
+		if err := r.helper.ProcessModuleSpec(ctrl.LoggerInto(ctx, logger), &nmc, &mod, statusMap[moduleNameKey]); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error while processing module %s: %v", moduleNameKey, err)
+		}
+
+		delete(statusMap, moduleNameKey)
+	}
+
+	// We have processed all module specs.
+	// Now, go through the remaining, "orphan" statuses that do not have a corresponding spec; those must be unloaded.
+
+	for statusNameKey, status := range statusMap {
+		logger := logger.WithValues("status", statusNameKey)
+
+		if err := r.helper.ProcessOrphanModuleStatus(ctrl.LoggerInto(ctx, logger), &nmc, status); err != nil {
+			return ctrl.Result{}, fmt.Errorf("could not create the unloader Pod for %s: %v", statusNameKey, err)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *NodeModulesConfigReconciler) SetupWithManager(ctx context.Context, mgr manager.Manager) error {
+	err := mgr.GetCache().IndexField(ctx, &v1.Pod{}, ".spec.nodeName", func(o client.Object) []string {
+		return []string{o.(*v1.Pod).Spec.NodeName}
+	})
+	if err != nil {
+		return fmt.Errorf("could not start the worker Pod indexer: %v", err)
+	}
+
+	nodeToNMCMapFunc := func(_ context.Context, o client.Object) []reconcile.Request {
+		return []reconcile.Request{
+			{NamespacedName: types.NamespacedName{Name: o.GetName()}},
+		}
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(NodeModulesConfigReconcilerName).
+		For(&kmmv1beta1.NodeModulesConfig{}).
+		Owns(
+			&v1.Pod{},
+			builder.WithPredicates(filter.HasLabel(actionLabelKey)),
```

Comment:
why do we need to filter on actionLabel? Isn't owner enough?

Reply:
This is probably slightly cheaper since it is just a map lookup.

Reply:
but you already have Owns... it means that it includes checking the owner reference, so it will be done anyway

Reply:
Well then indeed we can drop the predicate entirely.

Reply:
Removed.